uast/libuast/api.go

```diff
@@ -0,0 +1,278 @@
+package main
+
+/*
+#include "uast.h"
+#include <stdlib.h>
+*/
+import "C"
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"unsafe"
+
+	"gopkg.in/bblfsh/sdk.v2/uast/nodes"
+	"gopkg.in/bblfsh/sdk.v2/uast/nodes/nodesproto"
+	"gopkg.in/bblfsh/sdk.v2/uast/query"
+	"gopkg.in/bblfsh/sdk.v2/uast/yaml"
+)
+
+func main() {}
+
+func newUast(iface *C.NodeIface, h Handle) *C.Uast {
+	sz := unsafe.Sizeof(C.Uast{})
+	u := (*C.Uast)(C.malloc(C.size_t(sz)))
+	u.iface = iface
+	u.handle = C.uint64_t(h)
+	u.root = 0
+	u.ctx = 0
+	return u
+}
+
+//export UastNew
+// UastNew initializes a new UAST context that will use a provided node interface as an implementation.
+// This allows libuast to work with every language's native node data structures. Client can pass
+// an additional UastHandle to distinguish between different UAST context instances.
+//
+// The returned context pointer is guaranteed to be not NULL. Client should check LastError before
+// using the context and deallocate it with UastFree in case of an error occurs, or when the context
+// is no longer needed.
+func UastNew(iface *C.NodeIface, ctx C.UastHandle) *C.Uast {
+	cn := &cNodes{
+		impl: iface,
+	}
+	c := newContext(cn)
+
+	u := newUast(iface, c.Handle())
+	u.ctx = ctx
+	cn.ctx = u
+	return u
+}
+
+//export UastDecode
+// UastDecode accepts a pointer to a buffer with a specified size and decodes the content into
+// a new UAST structure.
+//
+// The new UAST context will use internal node interface implementation and all the nodes will
+// be managed by libuast.
+//
+// The returned context pointer is guaranteed to be not NULL. Client should check LastError before
```

Comment:
The main reason for this is that C is now only an API for the Go code and the Go code can be concurrent. Because of this I intentionally omitted any methods that touch a global state that is not bound to a specific object instance.